--- libcork/src/libcork/posix/env.c
+++ libcork/src/libcork/posix/env.c
@@ -17,13 +17,20 @@
 #include "libcork/helpers/errors.h"

 #if defined(__APPLE__)
-/* Apple doesn't provide access to the "environ" variable from a shared library.
- * There's a workaround function to grab the environ pointer described at [1].
- *
- * [1] http://developer.apple.com/library/mac/#documentation/Darwin/Reference/ManPages/man7/environ.7.html
- */
-#include <crt_externs.h>
-#define environ  (*_NSGetEnviron())
+
+#include <TargetConditionals.h>
+
+#if TARGET_OS_IPHONE
+    #define NO_ENVIRON 1
+#else
+    /* Apple doesn't provide access to the "environ" variable from a shared library.
+     * There's a workaround function to grab the environ pointer described at [1].
+     *
+     * [1] http://developer.apple.com/library/mac/#documentation/Darwin/Reference/ManPages/man7/environ.7.html
+     */
+    #include <crt_externs.h>
+    #define environ  (*_NSGetEnviron())
+#endif

 #else
 /* On all other POSIX platforms, we assume that environ is available in shared
@@ -93,6 +100,9 @@ cork_env_add_internal(struct cork_env *env, const char *name, const char *value)
 struct cork_env *
 cork_env_clone_current(void)
 {
+#ifdef NO_ENVIRON
+    return NULL;
+#else
     char  **curr;
     struct cork_env  *env = cork_env_new();

@@ -113,6 +123,7 @@ cork_env_clone_current(void)
     }

     return env;
+#endif
 }


@@ -192,7 +203,9 @@ cork_env_set_vars(void *user_data, struct cork_hash_table_entry *entry)
 static void
 clearenv(void)
 {
+#ifndef NO_ENVIRON
     *environ = NULL;
+#endif
 }

 #else
